Throw exception NotSupported Exception for `UuidGenerator` with doctr…

scyzoryck opened the pull request:
Let's fix another issue:
```
[0;31mERROR[0m: UndefinedMethod - lib/Doctrine/ORM/Id/UuidGenerator.php:33:59 - Method Doctrine\DBAL\Platforms\AbstractPlatform::getGuidExpression does not exist (see https://psalm.dev/022)
        $sql  = 'SELECT ' . $conn->getDatabasePlatform()->[97;41mgetGuidExpression[0m();
```
Method has been removed with https://github.com/doctrine/dbal/pull/3211. 
Lets throw exception if method do not exists. Generator itself is already deprecated. 

Questions:
1. May it would be better to throw it in constructor?
2. What about unit tests? Should we cover also this case? 

It looks like there will be few similar issues, so let's start from this one to create a template for the next PRs :)

Commit:
Throw exception NotSupported Exception for `UuidGenerator` with doctrine/dbal:3.x.

Generating `getGuidExpression` has been removed in doctrine/dbal:3.x.

Partially fixes #8884

greg0ire commented on lib/Doctrine/ORM/Id/UuidGenerator.php:
```diff
@@ -26,11 +29,17 @@ public function __construct()
     {
         $conn = $em->getConnection();
-        $sql  = 'SELECT ' . $conn->getDatabasePlatform()->getGuidExpression();
+        if (method_exists($conn, 'getGuidExpression')) {
```
This line looks wrong to me, for several reasons. If it is indeed wrong, I think that answers your question about unit tests :stuck_out_tongue:

greg0ire commented:
> May it would be better to throw it in constructor?

Throwing as early as possible is indeed a good idea 👍

scyzoryck commented:
Thanks @greg0ire! It was completly wrong. I've added Unit Tests to cover its logic. 
In PHPUnit I've defined 2 groups - dbal3 and dbal2 to differentiate tests that could be run with `doctrine/dbal` 2 and 3. I've adjusted also CI configuration.

greg0ire commented on psalm.xml:
```diff
@@ -47,5 +47,11 @@
+                <!-- See https://github.com/doctrine/orm/issues/8884 -->
+                <referencedMethod name="Doctrine\DBAL\Platforms\AbstractPlatform::getGuidExpression"/>
+            </errorLevel>
+        </UndefinedMethod>
```
Can you do the same for PHPStan? I can see an automatic comment above from that tool

scyzoryck replied:
@greg0ire Added.

beberlei commented on lib/Doctrine/ORM/Exception/NotSupported.php:
```diff
@@ -10,4 +10,9 @@ public static function create(): self
+
+    public static function createForDbal3(): self
+    {
+        return new self('This behaviour is not supported by doctrine/dbal:3.x, please use doctrine/dbal:2.x instead.');
```
I think we should improve this exception message, because it gives no indiication on how to get to DBAL 3 making it sound DBAL 2 is better. Maybe:

```suggestion
        return new self('Feature was deprecated in doctrine/dbal 2.x and is not supported by installed doctrine/dbal:3.x, please see the doctrine/deprecations logs for new alternative approaches.');
```

scyzoryck replied:
Thanks @beberlei! Sounds much better :)

beberlei reviewed (dismissed):
Small adjustment to the exception message, otherwise good to go for me.

greg0ire commented on phpstan.neon:
```diff
@@ -16,6 +16,8 @@ parameters:
         # https://github.com/doctrine/collections/pull/282
         - '/Variable \$offset in isset\(\) always exists and is not nullable\./'
+        # method removed in doctrine/dbal3.x
+        - '/^Call to an undefined method Doctrine\\DBAL\\Platforms\\AbstractPlatform::getGuidExpression\(\).$/'
```
This is only useful when using DBAL 3, and thus makes the DBAL 2 phpstan job fail :(
Maybe we should have a separate file for DBAL 3? Like `phpstan-dbal3.neon`?

greg0ire commented on phpstan-dbal3.neon:
```diff
@@ -0,0 +1,14 @@
+        - '/^Attribute class ReturnTypeWillChange does not exist.$/'
+
+        # https://github.com/doctrine/collections/pull/282
+        - '/Variable \$offset in isset\(\) always exists and is not nullable\./'
```
Shouldn't the 2 errors above be only in `phpstan.neon`? They only appear with DBAL 3? Or is `ignoreErrors` being overwritten here?

scyzoryck replied:
It looks like it might require a bit more work - as the ignored paths are merged from all files. May it would make sense to have 2 separate baselines and configs for dbal2 and dbal3?

greg0ire replied:
Maybe yes, but I'd rather do that when absolutely necessary.

scyzoryck replied:
@greg0ire - splitted a little bit more. Common params moved to separate file.  It looks like the best what we can do to split them in a logic way.

SenseException commented on .github/workflows/continuous-integration.yml:
```diff
@@ -40,14 +40,14 @@ jobs:
       - name: "Run PHPUnit"
         continue-on-error: true
-        run: "vendor/bin/phpunit -c ci/github/phpunit/sqlite.xml --coverage-clover=coverage-no-cache.xml"
+        run: "vendor/bin/phpunit -c ci/github/phpunit/sqlite.xml --coverage-clover=coverage-no-cache.xml --exclude-group performance,locking_functional,dbal2"
```
@greg0ire Are `performance` and `locking_functional` not needed for the regular "Run PHPUnit" step?

greg0ire replied:
`performance` is run separately, that's for sure. As for `locking_functional`, it's about Gearman, so it might be great to run it someday. They are already excluded in `sqlite.xml`, but I think @scyzroyck had to respecify them in order to add `dbal2` to the list.

SenseException replied:
We'll see how this will affects the code coverage in case some test won't be executed anymore.

greg0ire replied:
[looks like it does](https://github.com/doctrine/orm/pull/8898#discussion_r693210346)

greg0ire commented on lib/Doctrine/ORM/Id/UuidGenerator.php:
```diff
@@ -22,10 +26,16 @@ public function __construct()
+     * @throws NotSupported
      */
     public function generate(EntityManager $em, $entity)
     {
```
Apparently this method is no longer covered: https://app.codecov.io/gh/doctrine/orm/compare/8898/changes
Why is that?

scyzoryck replied:
If the coverage is taken from tests with doctrine/dbal:3.0, this function will not be tested as the class itself will be never initialised.

scyzoryck replied:
@greg0ire - it looks like the possible reason might be the same name `--coverage-clover=coverage-cache.xml` used for the tests with dbal2 and dbal3.

greg0ire replied:
It's close to that, but that's not it, I think you can have the same name as long as it differs when it's uploaded (so the upload step needs to be modified)

greg0ire replied:
Or maybe I'm mistaken, looks like I already took care of setting different names when introducing the DBAL 3 build

greg0ire replied:
> If the coverage is taken from tests with doctrine/dbal:3.0, this function will not be tested as the class itself will be never initialised.

The coverage is supposed to be merged from multiple reports

greg0ire replied:
The list can be seen here: https://github.com/doctrine/orm/pull/8898/checks?check_run_id=3343305825#step:3:21

I think some might be missing indeed.

greg0ire replied:
Ah but the DBAL3 tests don't even run right now, so there will not be a report from them. The DBAL2 tests are running though, and are supposed to cover that path.

greg0ire replied:
The job supposed to cover that path are the mysql ones, and they are present… not sure what to think :thinking:

SenseException replied:
Independent from this PR: Shouldn't this also be covered by the other platforms?

greg0ire replied:
:thinking: good question, might be worth removing that call to `self::markTestSkipped` and see on which other platforms it passes. The situation might have evolved since this test was written in 2012.

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/UUIDGeneratorTest.php:
```diff
@@ -20,18 +23,28 @@ class UUIDGeneratorTest extends OrmFunctionalTestCase
     use VerifyDeprecations;
 
+    /**
+     * @group dbal2
```
Why do we need these groups anyway? Can't the same be achieved with `markTestSkipped`?

greg0ire commented:
The mysql jobs are passing despite me pushing a very wrong commit, so it's not a coverage issue, it's an execution issue. Marking as draft until this is fixed by you or me :)

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/UUIDGeneratorTest.php:
```diff
@@ -20,18 +23,28 @@ class UUIDGeneratorTest extends OrmFunctionalTestCase
             self::markTestSkipped('Currently restricted to MySQL platform.');
         }
 
+        if (! method_exists(Connection::class, 'getGuidExpression')) {
```
I don't think `Connection` is the right class for this, I think you meant `AbstractPlatform`. That's probably why the code was not executed.

greg0ire replied:
Yup, that was it!

greg0ire commented:
Note: there is still a drop in coverage, because the DBAL 3 tests are not actually running yet (see #8886)

greg0ire commented:
Thanks @scyzoryck !
